pkg/proc/native/proc_freebsd.c

```diff
@@ -0,0 +1,93 @@
+#include <sys/param.h>
+#include <sys/mount.h>
+#include <sys/queue.h>
+#include <sys/sysctl.h>
+#include <sys/user.h>
+
+#include <libprocstat.h>
+#include <libutil.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "proc_freebsd.h"
+
+/*
+ * Returns the absolute pathname of the process's executable, if one was found.
+ * Must be freed by the caller. Sets errno on failure.
+ */
+char * find_executable(int pid) {
+	struct procstat *ps;
+	struct kinfo_proc *kp;
+	char *pathname;
+
+	/*
+	 * procstat_open_sysctl is for running processes. For core files, use
+	 * procstat_open_core
+	 */
+	ps = procstat_open_sysctl();
+	kp = kinfo_getproc(pid);
+	pathname = malloc(MNAMELEN);
+	if (ps && kp && pathname)
+		procstat_getpathname(ps, kp, pathname, MNAMELEN);
+	free(kp);
+	procstat_close(ps);
+	return (pathname);
+}
+
+/*
+ * Returns the comm value of the process, which is usually the basename of its
+ * executable. Must be freed by the caller.  Sets errno on failure.
+ */
+char * find_command_name(int pid) {
+	char *command_name = NULL;
+	struct kinfo_proc *kinfo;
+
+	kinfo = kinfo_getproc(pid);
+	if (kinfo != NULL) {
+		command_name = malloc(COMMLEN + 1);
+		if (command_name != NULL)
+			strlcpy(command_name, kinfo->ki_comm, COMMLEN + 1);
+		free(kinfo);
+	}
+
+	return (command_name);
+}
+
+int find_status(int pid){
```

Comment:
Looking at the [procfs manpage](https://www.freebsd.org/cgi/man.cgi?query=procfs&apropos=0&sektion=0&manpath=FreeBSD+10.4-RELEASE&arch=default&format=html) it seems to me that this and find_executable could be implemented using that. Am I wrong?

Reply:
It is not recommended to use procfs in FreeBSD as it requires some additional explicit changes in /etc/fstab (https://www.freebsd.org/doc/en/articles/linux-users/procfs.html), not all users have this Linux-compatibility option enabled. Using procstat calls is a recommended way to get info.

Reply:
Is it possible to do this without cgo?

Reply:
I believe no (except procfs which requires additional explicit mount).